File: 백준/Gold/1759. 암호 만들기/암호 만들기.swift
import Foundation

let input = readLine()!.split(separator:" ").map { Int($0)! }
let l = input[0], c = input[1]
// 최소 1개의 모음과 2개의 자음, 알파벳순 정렬
let letters = readLine()!.split(separator:" ").map { String($0) }.sorted()
var visited = Array(repeating: false, count: c)
let moeum = ["a", "e", "i", "o", "u"]
var answer:[String] = []

func isValidLetter(_ str: [String]) -> Bool {
  var moeum_count = 0
  var jaeum_count = 0
  
  for i in str {
    if moeum.contains(i) {
      moeum_count += 1
    } else {
      jaeum_count += 1
    }
  }

  return moeum_count >= 1 && jaeum_count >= 2
}

func backtracking(_ start: Int) {
  if answer.count == l {
    if isValidLetter(answer) {
      print(answer.joined())
    }
    return
  }

  for i in start..<c {
    if !visited[i] {
      visited[i] = true
        answer.append(letters[i])
       backtracking(i)
      answer.removeLast() 
      visited[i] = false
    }
  }
}

backtracking(0)

File: 백준/Gold/1759. 암호 만들기/암호 만들기.py
l, c = map(int, input().split())
arr = sorted(list(input().split()))
visited = [False] * c

def count_moeum_zaeum(str):
  moeums = ["a", "e", "i", "o", "u"]
  moeum = 0
  zaeum = 0
  
  for s in str:
    if s in moeums:
      moeum += 1
    else:
      zaeum += 1

  return moeum >= 1 and zaeum >= 2

def backtracking(str, idx):
  if count_moeum_zaeum(str) and len(str) == l:
    print(str)
    return

  for i in range(idx, c):
    if not visited[i]:
      visited[i] = True
      backtracking(str + arr[i], i)
      visited[i] = False

backtracking("", 0)